x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/AuthorizationUtils.java

```diff
@@ -150,6 +197,163 @@ public static void switchUserBasedOnActionOriginAndExecute(
+            return;
+        }
+
+        final String[] indices = indicesRequest.indices();
+        if (indices == null || indices.length == 0 || Arrays.equals(IndicesAndAliasesResolverField.NO_INDICES_OR_ALIASES_ARRAY, indices)) {
+            // No indices to check
+            return;
+        }
+
+        // Checks if the parent context has already successfully authorized access to the child's indices.
+        if (Arrays.stream(indices).allMatch(indicesAccessControl::hasIndexPermissions) == false) {
+            return;
+        }
+
+        try {
+            final Optional<ParentIndexActionAuthorization> existingParentAuthorization = Optional.ofNullable(
+                ParentIndexActionAuthorization.readFromThreadContext(threadContext)
+            );
+            if (existingParentAuthorization.isPresent()) {
+                if (existingParentAuthorization.get().action().equals(parentAction)) {
+                    // Single request can fan-out a child action to multiple nodes in the cluster, e.g. node1 and node2.
+                    // Sending a child action to node1 would have already put parent authorization in the thread context.
+                    // To avoid attempting to pre-authorize the same parent action twice we simply return here
+                    // since pre-authorization is already set in the context.
+                    logger.debug("child action [" + childAction + "] of parent action [" + parentAction + "] is already pre-authorized");
+                    return;
+                } else {
+                    throw new AssertionError(
+                        "Found pre-authorization for action ["
+                            + existingParentAuthorization.get().action()
+                            + "] while attempting to pre-authorize child action ["
+                            + childAction
+                            + "] of parent ["
+                            + parentAction
+                            + "]"
+                    );
+                }
+            } else {
+                logger.debug("adding pre-authorization for child action [" + childAction + "] of parent action [" + parentAction + "]");
+                new ParentIndexActionAuthorization(version, parentAction).writeToThreadContext(threadContext);
+            }
+        } catch (Exception e) {
+            logger.error("Failed to write authorization to thread context.", e);
+            throw internalError(
+                "Failed to write pre-authorization for child action ["
+                    + childAction
+                    + "] of parent action ["
+                    + parentAction
+                    + "] to the context"
+            );
+        }
+    }
+
+    @Nullable
+    public static AuthorizationEngine.AuthorizationContext extractAuthorizationContext(ThreadContext threadContext, String childAction) {
+        final String originatingAction = threadContext.getTransient(ORIGINATING_ACTION_KEY);
```

Comment:
>Conceptually, what is our intention here? Do we want "parent" or "originating"? I think "parent" is likely better. But we don't currently store the parent action. Is it part of the reason for having the parent/child lookup map? Would it be viable to actually store the parent action in threadContext?

Good question. We want "parent". I reused `ORIGINATING_ACTION` because I thought it's actually storing a parent action and wanted to avoid storing duplicate information. I was wrong.
The intention of parent/child lookup map was to be explicit which actions are allowed to be pre-authorized and as a side-effect was mitigating this problem with originating action.

Reply:
You are right. It does not solve the issue. Using originating action is wrong her. I'll try to describe the algorithm I have in mind (please correct me if my reasoning is wrong):

At the moment, we only want to pre-authorize search child actions. Idea is to have some lookup mechanism (it does not have to be a map) which allows us to check if the action should be pre-authorized or not. We want to be explicit in which actions we allow. If the action is not in this map, we do authorization regularly.

We want to store the "parent" action in thread context (as transient) and only set it as persistent before we send a request to a remote node in the same cluster. Receiving node should check the presence of this header and do the same check against the lookup map, before proceeding with the authorization.

Reply:
I think that sounds about right. Just a brainstorming idea: have you considered storing the entire chain of actions? It might be more flexible and easier to reason with than having two individual headers for "originating" and "parent"?

Reply:
Actually, Albert had the similar idea as well, but we discussed it in a different context and maybe ruled it out too soon.
I'll re-consider it again. Thanks.